django/thunderstore/repository/api/experimental/views/package_index.py

```diff
@@ -0,0 +1,81 @@
+from io import BytesIO
+from typing import List
+
+from django.db.models import F
+from django.shortcuts import redirect
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import serializers
+from rest_framework.exceptions import APIException
+from rest_framework.renderers import JSONRenderer
+from rest_framework.views import APIView
+from sentry_sdk import capture_exception
+
+from thunderstore.repository.models import (
+    APIExperimentalPackageIndexCache,
+    PackageVersion,
+    PackageVersionQuerySet,
+)
+
+
+class ServiceUnavailable(APIException):
+    status_code = 503
+    default_detail = "Service temporarily unavailable, try again later."
+    default_code = "service_unavailable"
+
+
+class PackageIndexEntry(serializers.Serializer):
+    # Many fields have been excluded from this serializer as a design choice
+    # to discourage API client design that's not future proof.
+
+    namespace = serializers.CharField()
+    name = serializers.CharField()
+    version_number = serializers.CharField()
+
+    file_format = serializers.CharField(source="format_spec")
+    file_size = serializers.IntegerField()
+
+    dependencies = serializers.SerializerMethodField()
+
+    def get_dependencies(self, instance: PackageVersion) -> List[str]:
+        return [x.full_version_name for x in instance.dependencies.all()]
+
+
+def serialize_package_index() -> bytes:
+    versions: PackageVersionQuerySet = PackageVersion.objects.active().annotate(
```

Comment:
Immutability would be the ultimate goal yes. `is_active = False` is the equivalent of a delete as far as anything publicly facing is considered outside of integrity checks & related errors, which we need to support for various reasons (e.g. legal) regardless if we want to or not. We do probably need to change that flag to work to act as `this exists in the ecosystem but we won't be able to serve it` at some point, so really it'd just mean we know a specific package exists but it's been made unavailable through us.

The main reason we do a full run each time is not actually because of that, but rather because while using an append-only approach (which would be ideal) making schema changes becomes challenging, and we don't have good supporting logic for that at least yet. Long-term I would like to move towards append-only immutable data structures as much as possible though.

On a more conceptual level, all API endpoints are always going to have to be considered mutable, but they can link to immutable resources. This API could be implemented as a chain of immutable documents that each link to the previous, and then have this API link to the latest one for example. Doing that in a way that actually makes sense is more complex though, since you probably don't want to serve each individual entry as it's own document. Attempt to solve that and you soon find yourself needing to differentiate between `content` and `representation`, as you can serialize & paginate & break down the same content in multiple different ways which all lead to the same end result when put back together.

TL;DR: Immutability is hard

Nice job paying attention to this 👍 seems like my various ramblings about immutability haven't been at least entirely wasted 😄